pep-0727.rst

```diff
@@ -0,0 +1,516 @@
+``list`` would be considered to be "``A list of user names``".
+
+Implementers would only be required to support the top level use case, where the
+documentation for ``to`` is considered to be "``Who to say hi to``".
+They could optionally support having conditional documentation for when the type
+of the parameter passed is of one type or another, but they are not required to do so.
+
+
+Duplication
+-----------
+
+If ``doc()`` is used multiple times in a single ``Annotated``, it would be
+considered invalid usage from the developer, for example:
+
+.. code-block::
+
+    def hi(
+      to: Annotated[str, doc("A user name"), doc("The current user name")],
+    ) -> None: ...
+
+
+Implementers can consider this invalid and are not required to support this to be
+considered conformant.
+
+Nevertheless, as it might be difficult to enforce it on developers, implementers
+can opt to support one of the ``doc()`` declarations.
+
+In that case, the suggestion would be to support the last one, just because
+this would support overriding, for example, in:
+
+.. code-block::
+
+    User = Annotated[str, doc("A user name")]
+
+    CurrentUser = Annotated[User, doc("The current user name")]
+
+
+Internally, in Python, ``CurrentUser`` here is equivalent to:
+
+.. code-block::
+
+    CurrentUser = Annotated[str, doc("A user name"), doc("The current user name")]
+
+
+For an implementation that supports the last ``doc()`` appearance, the above
+example would be equivalent to:
+
+.. code-block::
+
+    def hi(
+      to: Annotated[str, doc("The current user name")],
+    ) -> None: ...
+
+
+Early Adopters and Older Python Versions
+========================================
+
+For older versions of Python and early adopters of this proposal, ``doc()`` and
+``DocInfo`` can be imported from the ``typing_extensions`` package.
+
+.. code-block::
+
+    from typing import Annotated
+
+    from typing_extensions import doc
+
+
+    def hi(
+      to: Annotated[str, doc("The current user name")],
+    ) -> None: ...
+
+
+Rejected Ideas
+==============
+
+
+Standardize Current Docstrings
+------------------------------
+
+A possible alternative would be to support and try to push as a standard one of the
+existing docstring formats. But that would only solve the standardization.
+
+It wouldn't solve any of the other problems, like getting editor support
+(syntax checks) for library authors, the distance and duplication of information
+between a parameter definition and its documentation in the docstring, etc.
+
+
+Extra Metadata and Decorator
+----------------------------
+
+An earlier version of this proposal included several parameters to indicate whether
+an object is discouraged from use, what exceptions it may raise, etc.
+To allow also deprecating functions and classes, it was also expected
+that ``doc()`` could be used as a decorator. But this functionality is covered
+by ``typing.deprecated()`` in :pep:`702`, so it was dropped from this proposal.
```

Comment:
This is the first version of the proposal that's a PEP and that most people have read -- perhaps rephrase in terms of "this could be much wider, but we decided against it"